proxy/src/main/java/com/wavefront/agent/preprocessor/SpanSanitizeTransformer.java

```diff
@@ -18,59 +19,79 @@ public SpanSanitizeTransformer(final PreprocessorRuleMetrics ruleMetrics) {
     this.ruleMetrics = ruleMetrics;
   }
 
-  @Nullable
   @Override
-  public Span apply(@Nullable Span span) {
+  public Span apply(@Nonnull Span span) {
     long startNanos = ruleMetrics.ruleStart();
     boolean ruleApplied = false;
-    if (!charactersAreValid(span.getSource())) {
-      span.setSource(sanitize(span.getSource()));
-      ruleApplied = true;
+
+    // sanitize name and replace '*' with '-'
+    String name = span.getName();
+    if (name != null) {
+      span.setName(sanitizeValue(name).replace('*', '-'));
+      if (span.getName().equals(name)) {
+        ruleApplied = true;
+      }
+    }
+
+    // sanitize source
+    String source = span.getSource();
+    if (source != null) {
+      span.setSource(sanitize(source));
+      if (!ruleApplied && !span.getSource().equals(source)) {
+        ruleApplied = true;
+      }
     }
+
     if (span.getAnnotations() != null) {
       for (Annotation a : span.getAnnotations()) {
-        if (!charactersAreValid(a.getKey())) {
-          a.setKey(sanitize(a.getKey()));
-          ruleApplied = true;
+        // sanitize tag key
+        String key = a.getKey();
+        if (key != null) {
+          a.setKey(sanitize(key));
+          if (!ruleApplied && !a.getKey().equals(key)) {
+            ruleApplied = true;
+          }
+        }
+
+        // sanitize tag value
+        String value = a.getValue();
+        if (value != null) {
+          a.setValue(sanitizeValue(value));
+          if (!ruleApplied && !a.getValue().equals(value)) {
+            ruleApplied = true;
+          }
         }
       }
     }
+
     if (ruleApplied) {
       ruleMetrics.incrementRuleAppliedCounter();
     }
     ruleMetrics.ruleEnd(startNanos);
     return span;
   }
 
-  private boolean charactersAreValid(String s) {
-    if (s == null) {
-      return true;
-    }
-    for (int i = 0; i < s.length(); i++) {
-      if (!characterIsValid(s.charAt(i))) {
-        return false;
-      }
-    }
-    return true;
-  }
-
-  private boolean characterIsValid(char c) {
-    // Legal characters are 44-57 (,-./ and numbers), 65-90 (upper), 97-122 (lower), 95 (_)
-    return (44 <= c && c <= 57) || (65 <= c && c <= 90) || (97 <= c && c <= 122) || c == 95;
-  }
-
   /**
    * Sanitize a string so that every invalid character is replaced with a dash.
    */
   private String sanitize(String s) {
```

Comment:
Do you want to make Java SDK as source of truth for sanitize() and sanitizeValue() methods?

Reply:
Ideally yes, will need to open a PR on the SDK first because sanitize() currently has behavior that pertains only to deltas while sanitizeValue() wraps the string in quotes.

Reply:
Updated to use `sanitizeWithoutQuotes` from the latest version of the SDK